source/adapters/level_zero/v2/memory.cpp

```diff
@@ -305,30 +426,81 @@ ur_result_t urMemBufferPartition(ur_mem_handle_t hBuffer, ur_mem_flags_t flags,
                                  ur_buffer_create_type_t bufferCreateType,
                                  const ur_buffer_region_t *pRegion,
                                  ur_mem_handle_t *phMem) {
-  std::ignore = hBuffer;
-  std::ignore = flags;
-  std::ignore = bufferCreateType;
-  std::ignore = pRegion;
-  std::ignore = phMem;
-  logger::error("{} function not implemented!", __FUNCTION__);
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+  UR_ASSERT(bufferCreateType == UR_BUFFER_CREATE_TYPE_REGION,
+            UR_RESULT_ERROR_INVALID_ENUMERATION);
+  UR_ASSERT((pRegion->origin < hBuffer->getSize() &&
+             pRegion->size <= hBuffer->getSize()),
+            UR_RESULT_ERROR_INVALID_BUFFER_SIZE);
+
+  auto accessMode = getDeviceAccessMode(flags);
+
+  UR_ASSERT(isAccessCompatible(accessMode, hBuffer->getDeviceAccessMode()),
+            UR_RESULT_ERROR_INVALID_VALUE);
+
+  *phMem = new ur_mem_sub_buffer_t(hBuffer, pRegion->origin, pRegion->size,
+                                   accessMode);
+
+  return UR_RESULT_SUCCESS;
 }
 
 ur_result_t urMemBufferCreateWithNativeHandle(
     ur_native_handle_t hNativeMem, ur_context_handle_t hContext,
     const ur_mem_native_properties_t *pProperties, ur_mem_handle_t *phMem) {
-  std::ignore = hNativeMem;
-  std::ignore = hContext;
-  std::ignore = pProperties;
-  std::ignore = phMem;
-  logger::error("{} function not implemented!", __FUNCTION__);
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+
+  auto ptr = reinterpret_cast<void *>(hNativeMem);
+  bool ownNativeHandle = pProperties ? pProperties->isNativeHandleOwned : false;
+
+  // Get base of the allocation
+  void *base;
+  size_t size;
+  ZE2UR_CALL(zeMemGetAddressRange,
+             (hContext->getZeHandle(), ptr, &base, &size));
+  UR_ASSERT(ptr == base, UR_RESULT_ERROR_INVALID_VALUE);
+
+  ze_device_handle_t zeDevice;
+  ZeStruct<ze_memory_allocation_properties_t> memoryAttrs;
+  UR_CALL(
+      getMemoryAttrs(hContext->getZeHandle(), ptr, &zeDevice, &memoryAttrs));
+
+  if (memoryAttrs.type == ZE_MEMORY_TYPE_UNKNOWN) {
+    return UR_RESULT_ERROR_INVALID_VALUE;
+  }
+
+  ur_device_handle_t hDevice{};
+  if (zeDevice) {
+    hDevice = hContext->getPlatform()->getDeviceFromNativeHandle(zeDevice);
+    UR_ASSERT(hContext->isValidDevice(hDevice),
+              UR_RESULT_ERROR_INVALID_CONTEXT);
+  }
+
+  // assume read-write
+  auto accessMode = ur_mem_handle_t_::device_access_mode_t::read_write;
+
+  // TODO: legacy adapter copies back the content to the native ptr on buffer
```

Comment:
Yes, because you want the content's of the native buffer, after it has been destroyed, to reflect whatever operations were performed on the wrapped UR buffer.

This is an actual use-case in SYCL. One of the examples in the book actually expect this behavior.

Reply:
Ok, done. I also added a test for that.